sdk/core/azure-core-http-vertx/src/main/java/com/azure/core/http/vertx/VertxAsyncHttpClient.java

```diff
@@ -0,0 +1,122 @@
+     */
+    VertxAsyncHttpClient(io.vertx.core.http.HttpClient client, Vertx vertx) {
+        Objects.requireNonNull(client, "client cannot be null");
+        Objects.requireNonNull(vertx, "vertx cannot be null");
+        this.client = client;
+        this.scheduler = Schedulers.fromExecutor(vertx.nettyEventLoopGroup());
+    }
+
+    @Override
+    public Mono<HttpResponse> send(HttpRequest request) {
+        return send(request, Context.NONE);
+    }
+
+    @Override
+    public Mono<HttpResponse> send(HttpRequest request, Context context) {
+        boolean eagerlyReadResponse = (boolean) context.getData("azure-eagerly-read-response").orElse(false);
+        return Mono.create(sink ->
+            toVertxHttpRequest(request).subscribe(vertxHttpRequest -> {
+                vertxHttpRequest.exceptionHandler(sink::error);
+
+                HttpHeaders requestHeaders = request.getHeaders();
+                if (requestHeaders != null) {
+                    requestHeaders.stream().forEach(header -> vertxHttpRequest.putHeader(header.getName(), header.getValuesList()));
+                    if (request.getHeaders().get("Content-Length") == null) {
+                        vertxHttpRequest.setChunked(true);
+                    }
+                } else {
+                    vertxHttpRequest.setChunked(true);
+                }
+
+                vertxHttpRequest.response(event -> {
+                    if (event.succeeded()) {
+                        HttpClientResponse vertxHttpResponse = event.result();
+                        vertxHttpResponse.exceptionHandler(sink::error);
+
+                        if (eagerlyReadResponse) {
+                            vertxHttpResponse.body(bodyEvent -> {
+                                if (bodyEvent.succeeded()) {
+                                    sink.success(new BufferedVertxHttpResponse(request, vertxHttpResponse, bodyEvent.result()));
+                                } else {
+                                    sink.error(bodyEvent.cause());
+                                }
+                            });
+                        } else {
+                            sink.success(new VertxHttpAsyncResponse(request, vertxHttpResponse));
+                        }
+                    }
+                });
+
+                getRequestBody(request).subscribeOn(scheduler).subscribe(buffer -> {
+                    while (buffer.hasRemaining()) {
+                        int bufferSize;
+                        if (buffer.remaining() > BYTE_BUFFER_CHUNK_SIZE) {
+                            bufferSize = BYTE_BUFFER_CHUNK_SIZE;
+                        } else {
+                            bufferSize = buffer.remaining();
+                        }
+
+                        byte[] bytes = new byte[bufferSize];
+                        buffer.get(bytes, 0, bytes.length);
+
+                        vertxHttpRequest.write(Buffer.buffer(bytes));
+                    }
```

Comment:
Reading through how `Buffer.buffer(byte[])` works this appears that'll copy the `ByteBuffer` data at minimum twice. Creating `bytes` is the first copy, calling `Buffer.buffer(byte[])` will be the second copy as based on the documentation this is a copy constructor: https://vertx.io/docs/apidocs/io/vertx/core/buffer/Buffer.html#buffer-byte:A-

Does the `Buffer` being written need to be buffered data? If not, could we look into using Netty's ByteBuf API here: https://vertx.io/docs/apidocs/io/vertx/core/buffer/Buffer.html#buffer-io.netty.buffer.ByteBuf-

And create Unpooled ByteBufs that are backed by the `ByteBuffer`? https://netty.io/4.1/api/io/netty/buffer/Unpooled.html#wrappedBuffer-java.nio.ByteBuffer-

Reply:
I'm struggling a bit with this. Would it be enough to do something like the following? Or is something more complicated needed?

```java
getRequestBody(request)
    .map(Unpooled::wrappedBuffer)
    .map(Buffer::buffer)
    .subscribe(buffer -> vertxHttpRequest.write(buffer), sink::error, vertxHttpRequest::end);
```

Reply:
Yeah, that's what I was thinking!

Reply:
Thanks for clarifying! I've made & pushed that change.